### PR TITLE
Payments and invoices filtering by creation date

### PR DESCRIPTION
**Pull request by yyforyongyu**
Replaces #6544 and #6542. Updates from the original PRs
- the validation of datetime params are placed in rpc server, not in `channeldb`
- update cli to use the new fields

### CODE REVIEW

File: lnrpc/lightning.proto [hunk 1]
```diff
@@ -3633,7 +3633,16 @@ message ListInvoiceRequest {
     specified index offset. This can be used to paginate backwards.
     */
     bool reversed = 6;
+
+    // If set, filters out all invoices with a creation date greater than or
+    // euqal to it. Measured in seconds since the unix epoch.
+    uint64 creation_date_start = 7;
+
+    // If set, filters out all invoices with a creation date less than  or
```
Comment (guggero) [inline lnrpc/lightning.proto:3641]: nit: double space.

File: channeldb/invoice_test.go [hunk 2]
```diff
@@ -1090,6 +1090,8 @@ func TestQueryInvoices(t *testing.T) {
 	for i := 1; i <= numInvoices; i++ {
 		amt := lnwire.MilliSatoshi(i)
 		invoice, err := randInvoice(amt)
+		invoice.CreationDate = invoice.CreationDate.Add(
+			time.Duration((i - 1)) * time.Second)
```
Comment (guggero) [inline channeldb/invoice_test.go:1094]: nit: formatting and superfluous inner parenthesis.

File: channeldb/invoices.go [hunk 3]
```diff
@@ -1349,6 +1361,23 @@ func (d *DB) QueryInvoices(q InvoiceQuery) (InvoiceSlice, error) {
...
+				q.CreationDateStart,
+			) {
+
+				return false, nil
+			}
+
+			// Skip any invoices that were created after the
+			// specified time.
+			if endDateSet && invoice.CreationDate.After(
+				q.CreationDateEnd) {
```
Comment (guggero) [inline channeldb/invoices.go:1376]: nit: formatting.

File: rpcserver.go [hunk 4]
```diff
@@ -5516,6 +5516,16 @@ func (r *rpcServer) ListInvoices(ctx context.Context,
 		req.NumMaxInvoices = 100
 	}
 
+	// If both dates are set, we check that the start date is less than the
+	// end date, otherwise we'll get an empty result.
+	if req.CreationDateStart != 0 && req.CreationDateEnd != 0 {
+		if req.CreationDateStart >= req.CreationDateEnd {
+			return nil, fmt.Errorf("start date(%v) must be less "+
```
Comment (guggero) [inline rpcserver.go:5523]: nit: s/less than/before/?

File: lnrpc/lightning.proto [hunk 5]
```diff
@@ -3841,6 +3841,14 @@ message ListPaymentsRequest {
     of payments, as all of them have to be iterated through to be counted.
     */
     bool count_total_payments = 5;
+
+    // If set, filters out all invoices with a creation date greater than or
+    // euqal to it. Measured in seconds since the unix epoch.
+    uint64 creation_date_start = 6;
+
+    // If set, filters out all invoices with a creation date less than  or
```
Comment (guggero) [inline lnrpc/lightning.proto:3849]: nit: same double space here.

File: channeldb/payments.go [hunk 6]
```diff
@@ -615,6 +627,23 @@ func (d *DB) QueryPayments(query PaymentsQuery) (PaymentsResponse, error) {
...
+				query.CreationDateStart,
+			) {
+
+				return false, nil
+			}
+
+			// Skip any payments that were created after the
+			// specified time.
+			if endDateSet && payment.Info.CreationTime.After(
+				query.CreationDateEnd) {
```
Comment (guggero) [inline channeldb/payments.go:642]: nit: same formatting.

File: channeldb/payments_test.go [hunk 7]
```diff
@@ -426,6 +468,9 @@ func TestQueryPayments(t *testing.T) {
 					t.Fatalf("unable to create test "+
 						"payment: %v", err)
 				}
+				// Override creation time to allow for testing
+				// of MinCreationTimeNs and MaxCreationTimeNs.
```
Comment (guggero) [inline channeldb/payments_test.go:472]: nit: The fields are being called differently now.

**Review by guggero — APPROVED** [review]
Very nice, thanks for picking this up!
Only nits :tada:

File: channeldb/invoice_test.go (hunk 2 above)
Comment (yyforyongyu) [reply channeldb/invoice_test.go:1094]: hawkeye! fixed!

File: channeldb/invoices.go (hunk 3 above)
Comment (yyforyongyu) [reply channeldb/invoices.go:1376]: fixed!

File: channeldb/payments.go (hunk 6 above)
Comment (yyforyongyu) [reply channeldb/payments.go:642]: fixed!

File: channeldb/payments_test.go (hunk 7 above)
Comment (yyforyongyu) [reply channeldb/payments_test.go:472]: fixed!

File: lnrpc/lightning.proto (hunk 1 above)
Comment (yyforyongyu) [reply lnrpc/lightning.proto:3641]: fixed!

File: lnrpc/lightning.proto (hunk 5 above)
Comment (yyforyongyu) [reply lnrpc/lightning.proto:3849]: fixed!

File: rpcserver.go (hunk 4 above)
Comment (yyforyongyu) [reply rpcserver.go]: switched!

File: lnrpc/lightning.proto [hunk 8]
```diff
@@ -3633,7 +3633,16 @@ message ListInvoiceRequest {
     specified index offset. This can be used to paginate backwards.
     */
     bool reversed = 6;
+
+    // If set, filters out all invoices with a creation date greater than or
```
Comment (positiveblue) [inline lnrpc/lightning.proto:3637]: q: shouldn't this be "filters out all invoices with a creation date less than"?

File: lnrpc/lightning.proto [hunk 9]
```diff
@@ -3633,7 +3633,16 @@ message ListInvoiceRequest {
...
+    // euqal to it. Measured in seconds since the unix epoch.
+    uint64 creation_date_start = 7;
+
+    // If set, filters out all invoices with a creation date less than or euqal
```
Comment (positiveblue) [inline lnrpc/lightning.proto:3641]: q: shouldn't this be "filters out all invoices with a creation date greater than"?

File: cmd/lncli/cmd_invoice.go [hunk 10]
```diff
@@ -234,6 +234,18 @@ var listInvoicesCommand = cli.Command{
 			Usage: "if set, invoices succeeding the " +
 				"index_offset will be returned",
 		},
+		cli.Uint64Flag{
+			Name: "creation_date_start",
+			Usage: "timestamp in seconds, if set, filter " +
```
Comment (positiveblue) [inline cmd/lncli/cmd_invoice.go:239]: I think here is right because it says filter and in the other comment says "filter out".

**Review by positiveblue — APPROVED** [review]
LGTM @yyforyongyu  I just left a couple of comments about the params description.

I will make sure that this changes are incorporated in #7215

File: cmd/lncli/cmd_invoice.go (hunk 10 above)
Comment (yyforyongyu) [reply cmd/lncli/cmd_invoice.go:239]: hmm yeah changed the `filter out` to be more clear.

File: lnrpc/lightning.proto (hunk 8 above)
Comment (yyforyongyu) [reply lnrpc/lightning.proto:3637]: make sense. Used "return" instead to avoid confusion.

File: lnrpc/lightning.proto (hunk 9 above)
Comment (yyforyongyu) [reply lnrpc/lightning.proto:3641]: switched to "return"

File: lnrpc/lightning.proto [hunk 11]
```diff
@@ -3631,7 +3633,16 @@ message ListInvoiceRequest {
     specified index offset. This can be used to paginate backwards.
     */
     bool reversed = 6;
+
+    // If set, returns all invoices with a creation date greater than or euqal
```
Comment (alexbosworth) [inline lnrpc/lightning.proto:3637]: euqal = equal

Comment (yyforyongyu) [reply lnrpc/lightning.proto:3637]: oops good catch!

**Pull request merged**